feat(v4): z.creditCard() - support Credit Card Validation

ysknsid25 opened the pull request:
issue: https://github.com/colinhacks/zod/issues/4039

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Added robust credit card validation to string input checks, ensuring proper formatting and verification with clear error feedback.
- **Tests**
  - Integrated comprehensive tests verifying various valid and invalid credit card formats, ensuring reliable validation across multiple card types.
- **Documentation**
  - Updated README files to include the new `creditCard()` method in the `z.string()` schema for credit card validation.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

RobinTail commented on src/types.ts:
```diff
@@ -770,6 +807,37 @@ function isValidCidr(ip: string, version?: IpVersion) {
+    // Check if it matches a provider
+    creditByProviderRegexList.some((regex) => regex.test(sanitized!)) &&
+    // Check if passes luhn algorithm
+    isLuhnAlgo(sanitized)) as boolean;
```
already typed as a function returning `boolean`

```suggestion
    isLuhnAlgo(sanitized));
```

ysknsid25 committed:
feat: z.string.creditCard() - support Credit Card Validation

Signed-off-by: ysknsid25 <kengo071225@gmail.com>

ysknsid25 commented on src/types.ts:
```diff
@@ -770,6 +807,37 @@ function isValidCidr(ip: string, version?: IpVersion) {
 }
 
+function isValidCreditCard(cardNumber: string): boolean {
+  const isLuhnAlgo = (cardNumber: string): boolean => {
```
I renamed `cardNumber` to `sanitizedCardNumber` 🙋

RobinTail commented on packages/zod/src/v3/types.ts:
```diff
@@ -726,6 +775,39 @@ function isValidCidr(ip: string, version?: IpVersion) {
 
+function isValidCreditCard(cardNumber: string): boolean {
+  const isLuhnAlgo = (sanitizedCardNumber: string): boolean => {
+    // Remove any non-digit chars
+    const number = sanitizedCardNumber.replace(nonDigitRegex, "");
```
I suggest to combine it with the line `799` where you do sanitation:

```ts
const sanitized = cardNumber.replace(creditSanitizeRegex, "");
```

Otherwise, it seems that the argument is "sanitized", but not enough yet

ysknsid25 replied:
thnx. I fixed it!

ysknsid25 committed:
feat: z.string.creditCard() - support Credit Card Validation

Signed-off-by: ysknsid25 <kengo071225@gmail.com>

RobinTail reviewed (commented):
I believe it's a nice and useful feature, but as a recently [found out](https://github.com/colinhacks/zod/pull/4391#issuecomment-2921619876), v3 might no longer accept ones while being in a maintenance mode after releasing v4.
Perhaps it would worth it to move it into v4 as a `ZodCheck`, @ysknsid25 ?

ysknsid25 commented:
@RobinTail 
Should I implement it using ZodEmail as a reference? 👀
https://github.com/ysknsid25/zod/blob/ad2fc5ee221bc5fbfad1b7695c71a8a2c9c57314/packages/zod/src/v4/core/schemas.ts#L394-L408

ysknsid25 committed:
feat: z.string.creditCard() - support Credit Card Validation

Signed-off-by: ysknsid25 <kengo071225@gmail.com>

ysknsid25 commented:
@RobinTail 
It should be fixed so that it can be used with v4. Please check 🙏

ysknsid25 committed:
feat: z.string.creditCard() - support Credit Card Validation

Signed-off-by: ysknsid25 <kengo071225@gmail.com>

ysknsid25 committed:
Merge branch 'main' into feature/add-credit-card-validate

Signed-off-by: ysknsid25 <kengo071225@gmail.com>

ysknsid25 committed:
feat: v3/v4 z.string.creditCard()/z.creditCard() - support Credit Card Validation
add he.ts

Signed-off-by: ysknsid25 <kengo071225@gmail.com>

ysknsid25 committed:
Merge remote-tracking branch 'origin/main' into feature/add-credit-card-validate

# Conflicts:
#	packages/zod/src/v4/classic/schemas.ts

ysknsid25 committed:
feat: align z.creditCard()/z.string().creditCard() with upstream PR #5931

Replicate the implementation from colinhacks/zod#5931 verbatim onto this
branch so the two stay in sync. Compared with the prior in-branch version:

- regex sources moved to v4/core/regexes.ts (creditCard +
  creditCardProviders); v4/core/schemas.ts now consumes them
- Mastercard provider regex wrapped in a non-capturing group to fix the
  `|` precedence bug that let 13-digit Luhn-valid Mastercard-prefix
  numbers through (e.g. "5100000000003")
- outer-shape regex widened from `\d{14,19}` to `\d{13,19}` to cover
  legitimate 13-digit issuers
- _creditCard() now accepts $ZodCreditCardParams (was hard-coded params type)
- ZodMiniCreditCard rebased on _ZodMiniString<$ZodCreditCardInternals>
- credit_card label added to all 49 locale dictionaries
- dedicated credit-card.test.ts files for v3 and v4/classic; ad-hoc tests
  in string.test.ts/validations.test.ts reverted

ysknsid25 committed:
revert: drop v3 creditCard support

Signed-off-by: ysknsid25 <kengo071225@gmail.com>

RobinTail reviewed (approved):
🥇 
I like this feature.

ysknsid25 committed:
test: add 13 digit VISA test case

Signed-off-by: ysknsid25 <kengo071225@gmail.com>

ysknsid25 commented:
@RobinTail 

Sorry, I added test case. please re-review it 🙇
